controllers/elfmachine_controller.go

```diff
@@ -1160,3 +1168,104 @@ func (r *ElfMachineReconciler) getK8sNodeIP(ctx *context.MachineContext, nodeNam
 
 	return "", nil
 }
+
+// Handling duplicate virtual machines.
+//
+// NOTE: These will be removed when Tower fixes issue with duplicate virtual machines.
+
+const (
+	// duplicateVMDefaultRequeueTimeout is the default time for how long to wait when
+	// requeueing a duplicate VM operation.
+	duplicateVMDefaultRequeueTimeout = 1 * time.Minute
+
+	// waitDuplicateVMTaskInterval is the default interval time polling task.
+	waitDuplicateVMTaskInterval = 10 * time.Second
+
+	// waitDuplicateVMTaskTimeout is the default timeout for waiting for task to complete.
+	waitDuplicateVMTaskTimeout = 1 * time.Minute
+)
+
+// deleteDuplicateVMs deletes the duplicate virtual machines.
+// Only be used to delete duplicate VMs before the ElfCluster is deleted.
+func (r *ElfMachineReconciler) deleteDuplicateVMs(ctx *context.MachineContext) (reconcile.Result, error) {
+	vms, err := ctx.VMService.FindVMsByName(ctx.ElfMachine.Name)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if len(vms) <= 1 {
+		return reconcile.Result{}, nil
+	}
+
+	if ctx.ElfMachine.Status.VMRef == "" {
```

Comment:
1. VMRef会在创建过程中记录vm的ID(而不是LocalID)，这里只判断VMRef非空，有可能误删除正常VM吗？  
2. 另外 建议加上过滤掉带 label sks-managed 的VM，不做删除操作。

```
// GetVMRef returns the ID or localID of the VM.
// If the localID is in UUID format, return the localID, otherwise return the ID.
//
// Before the ELF VM is created, Tower sets a "placeholder-{UUID}" format string to localID, such as "placeholder-7d8b6df1-c623-4750-a771-3ba6b46995fa".
// After the ELF VM is created, Tower sets the VM ID in UUID format to localID.
func GetVMRef(vm *models.VM) string {
```

Reply:
标签需要再调一次接口。

Reply:
我担心有误删除的可能。这样吧，对于开机状态的同名VM(理论上不会出现此情况)，不要进行删除。